compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -889,6 +889,27 @@ private boolean isAssignable(BType source, BType target, Set<TypePair> unresolve
                 isArrayTypesAssignable((BArrayType) source, target, unresolvedTypes);
     }
 
+    private boolean isMutable(BType type) {
+        if (Symbols.isFlagOn(type.flags, Flags.READONLY)) {
+            return false;
+        }
+
+        if (type.tag != TypeTags.UNION) {
+            return true;
+        }
+
+        BUnionType unionType = (BUnionType) type;
+        for (BType memberType : unionType.getMemberTypes()) {
+            if (!Symbols.isFlagOn(memberType.flags, Flags.READONLY)) {
+                return true;
+            }
+        }
+
+        unionType.flags |= Flags.READONLY;
+        unionType.tsymbol.flags |= Flags.READONLY;
```

Comment:
Do we need to fix this later? I think union creation should properly add this flag.

Reply:
The union creation actually already handles setting the read-only flag if all the members have the read-only flag set. The issue here was due to implicitly read-only types - for such types the read-only flag may be set after the creation of the union type and thus the read-only flag may not be set even though all the members have it set.

I've created https://github.com/ballerina-platform/ballerina-lang/issues/28012 to track revisiting this.